src/gretel_trainer/benchmark/session.py

```diff
@@ -0,0 +1,374 @@
+            trainer_project_name = _trainer_project_name(
+                self._config, trainer_project_index
+            )
+            self._trainer_project_names[run_identifier] = trainer_project_name
+            return GretelTrainerStrategy(
+                benchmark_model=benchmark_model,
+                dataset=dataset,
+                run_identifier=run_identifier,
+                project_name=trainer_project_name,
+                config=self._config,
+            )
+        else:
+            return GretelSDKStrategy(
+                benchmark_model=benchmark_model,
+                dataset=dataset,
+                artifact_key=artifact_key,
+                run_identifier=run_identifier,
+                project=self._project,
+                config=self._config,
+            )
+
+
+def _run_gretel(executor: Executor) -> None:
+    executor.run()
+
+
+def _run_custom(executors: list[Executor]) -> None:
+    for executor in executors:
+        executor.run()
+
+
+def _validate_jobs(config: BenchmarkConfig, jobs: list[JobSpec]) -> None:
+    gretel_models = [j.model for j in jobs if isinstance(j.model, GretelModel)]
+    if config.trainer:
+        _validate_trainer_setup(gretel_models)
+    else:
+        _validate_sdk_setup(gretel_models)
+
+
+def _validate_trainer_setup(gretel_models: list[GretelModel]) -> None:
+    unsupported_models = []
+    for model in gretel_models:
+        if model.trainer_model_type is None:
+            logger.error(
+                f"Model `{model.name}` (model key `{model.model_key}`) is not supported by Trainer. "
+                "Either remove it from this comparison, or configure this comparison to use the SDK (trainer=False)"
+            )
+            unsupported_models.append(model)
+    if len(unsupported_models) > 0:
+        raise BenchmarkException("Invalid configuration")
+
+
+def _validate_sdk_setup(gretel_models: list[GretelModel]) -> None:
+    if any(isinstance(m, GretelAuto) for m in gretel_models):
+        logger.error(
+            "GretelAuto is only supported when using Trainer. "
+            "Either remove it from this comparison, or configure this comparison to use Trainer (trainer=True)"
+        )
+        raise BenchmarkException("Invalid configuration")
+
+
+def model_name(model: Union[GretelModel, CustomModel]) -> str:
+    if isinstance(model, GretelModel):
+        return model.name
+    else:
+        return type(model).__name__
+
+
+def _trainer_project_name(config: BenchmarkConfig, index: int) -> str:
+    prefix = config.project_display_name
+    name = f"{prefix}-{index}"
+    return name.replace("_", "-")
+
+
+def _upload_dataset_to_project(
+    source: str, project: Project, trainer: bool
+) -> Optional[str]:
+    if trainer:
+        return None
+
+    return project.upload_artifact(source)
+
+
+def _make_run_key(model: Union[GretelModel, CustomModel], dataset: Dataset) -> RunKey:
+    return RunKey((model_name(model), dataset.name))
```

Comment:
This is a bit outside the scope of this PR, and I might be missing something, but I noticed it seems like `RunKey` could now just be a property of the `JobSpec` dataclass instead of needing this builder function.

Reply:
Good idea!